app/controllers/api/v0/signatures_controller.rb

```diff
@@ -1,21 +1,38 @@
 class Api::V0::SignaturesController < ApiController
   def create
     manifest = find_manifest(params[:manifest_id])
-    sign_request = read_body_as_json
+    signature_request = prep_signature_request(manifest)
 
-    manifest_content = manifest[:content].to_json
-    sign_request[:manifest_content] = manifest_content
-    system_session_token = session[:system_session_token]
-    sign_request["token"] = system_session_token
+    cdx_start = Time.now
+    cdx_response = CDX::Manifest.new(signature_request).sign
+    cdx_stop = Time.now
+    Rails.logger.debug(ANSI.blue{ "  CDX signature time: #{sprintf('%#g', (cdx_stop - cdx_start))} seconds" })
 
-    response = CDX::Manifest.new(sign_request).sign
+    status = update_manifest(cdx_response, signature_request, manifest)
 
-    if (response.key?(:document_id))
-      manifest[:document_id] = response[:document_id]
-      manifest[:activity_id] = sign_request["activity_id"]
-      manifest.save!
+    render(json: cdx_response.to_json, status: status) unless performed?
+  end
+
+  private
+
+  def prep_signature_request(manifest)
+    signature_request = read_body_as_json(symbolize_names: true)
+    signature_request[:manifest_content] = manifest[:content].to_json
+    if signature_request[:token]
+      signature_request[:token] = Base64.strict_decode64(signature_request[:token])
     end
+    signature_request
+  end
 
-    render json: response.to_json unless performed?
+  def update_manifest(cdx_response, signature_request, manifest)
+    if cdx_response.key?(:document_id)
+      manifest.document_id = cdx_response[:document_id]
+      manifest.activity_id = signature_request[:activity_id]
+      manifest.save!
+      status = 200
+    else
+      status = 422
```

Comment:
argh sorry for lack of clarity, what I meant was I think you can do this:

```ruby
def update_manifest(cdx_response, signature_request, manifest)
  if cdx_response.key?(:document_id)
    manifest.document_id = cdx_response[:document_id]
    manifest.activity_id = signature_request[:activity_id]
    manifest.save!
    200
  else
    422
   end
end
```

and it will be equal to what you have here